2018/CreateAsyncFromSyncIterator.js

```diff
@@ -0,0 +1,183 @@
+'use strict';
+
+var internalSlot = require('internal-slot');
+var GetIntrinsic = require('../GetIntrinsic.js');
+
+var AsyncFromSyncIteratorContinuation = require('./AsyncFromSyncIteratorContinuation.js');
+var Call = require('./Call.js');
+var CreateIterResultObject = require('./CreateIterResultObject.js');
+var CreateMethodProperty = require('./CreateMethodProperty.js');
+var DefinePropertyOrThrow = require('./DefinePropertyOrThrow.js');
+var Get = require('./Get.js');
+var GetMethod = require('./GetMethod.js');
+var IteratorNext = require('./IteratorNext.js');
+var OrdinaryObjectCreate = require('./ObjectCreate.js');
+var Type = require('./Type.js');
+
+var $asyncIterator = GetIntrinsic('%Symbol.asyncIterator%', true);
+var $toStringTag = GetIntrinsic('%Symbol.toStringTag%', true);
+var $TypeError = GetIntrinsic('%TypeError%');
+
+// TODO: Use %AsyncIterator.from% once it's in ECMA-262
+
+var AsyncIteratorPrototype
+	= GetIntrinsic('%AsyncIteratorPrototype%', true)
+	|| (function () {
+		var result = {};
+		if ($toStringTag) {
+			DefinePropertyOrThrow(result, $toStringTag, {
+				'[[Writable]]': false,
+				'[[Enumerable]]': false,
+				'[[Configurable]]': true,
+				'[[Value]]': 'AsyncIterator'
+			});
+		}
+		if ($asyncIterator) {
+			var method = {
+				'[Symbol.asyncIterator]': function () {
+					return this;
+				}
+			}['[Symbol.asyncIterator]'];
+			CreateMethodProperty(result, $asyncIterator, method);
+		}
+		return result;
+	}());
+
+var AsyncFromSyncIteratorPrototype
+	= GetIntrinsic('%AsyncFromSyncIteratorPrototype%', true)
+	// eslint-disable-next-line max-lines-per-function
+	|| (function () {
+		var $Promise = GetIntrinsic('%Promise%', true);
+		if (!$Promise) {
+			return;
+		}
+
+		// eslint-disable-next-line no-shadow
+		var AsyncFromSyncIteratorPrototype = OrdinaryObjectCreate(AsyncIteratorPrototype);
+		CreateMethodProperty(AsyncFromSyncIteratorPrototype, 'next', function next(value) {
```

Comment:
Arguably, these function bodies should be wrapped in the callback to `new $Promise`.